facets_atlasmaker/parallelize.py

```diff
@@ -0,0 +1,110 @@
+                            supported on old versions on PIL, just pillow.
+    disk_cache: Store intermediary image objects to disk. Not supported yet.
+
+  Returns:
+    A tuple (Image object or None if fails, status message string). Status
+    message string will be empty if success, or error message if failure.
+
+  Exceptions handled:
+    IOError: error retrieving image file, or truncated image file.
+    DecompressionBombError: Image is too large (>0.5G). See PIL documentation
+                            for instructions on setting a higher threshold.
+  """
+  if disk_cache:
+    raise NotImplementedError()
+
+  if allow_truncated_images:
+    try:
+      ImageFile.LOAD_TRUNCATED_IMAGES = True
+    except AttributeError as e:
+      logging.warning('Are you using PILLOW and not a very old version of PIL? '
+                      'Unable to force load of truncated image files: %s', e)
+
+  try:
+    src_image = atlasmaker_io.get_image(image_location)
+  except (IOError, DecompressionBombError) as e:
+    logging.error('Retrieval of file %s failed with error: %s',
+                  image_location, e)
+    return None, str(e)
+
+  try:
+    image_converter = convert.ImageConverter(src_image, image_convert_settings)
+    logging.debug('Successfully converted image: %s' % image_location)
+    return image_converter.convert(), ''
+  except IOError as e:
+    logging.error('Conversion of file %s failed with error: %s',
+                  image_location, e)
+    return None, str(e)
+
+
+def get_and_convert_images_parallel(image_src_locations, image_convert_settings,
+                                    n_jobs=-1, disk_cache=False, threads=False,
+                                    verbose=10, allow_truncated_images=False):
+  """Parallelize retrieving and converting image tasks.
+
+  Args:
+    images: List of source image paths (filepaths, URLs, etc).
+    image_convert_settings: ImageConvertSettings object.
+    disk_cache:: If True, will cache converted images to disk.
+    threads: If true, use threads instead of processes.
+    verbose: verbosity level for parallel. See joblib.Parallel documentation.
+    allow_truncated_images: If True, PIL will be tolerant of truncated image
+                            files and load/process them. Note that this isn't
+                            supported on old versions on PIL, just pillow.
+
+  Returns:
+    List of tuples, where each tuple contains
+    (converted Image object or None, status/error message string).
+  """
+  logging.info('Parallelizing with setting %d jobs' % n_jobs)
+
+  backend = None
+  if threads:
+    logging.debug('Parallelizing using threads.')
+    backend = 'threading'
+
+  outputs = Parallel(n_jobs=n_jobs, backend=backend, verbose=verbose)(
+      delayed(get_and_convert_image)(
+          location, image_convert_settings,
+          allow_truncated_images=allow_truncated_images,
+          disk_cache=disk_cache)
+      for location in image_src_locations)
```

Comment:
Currently, the output of Parallel looks something like this:

```
[Parallel(n_jobs=-1)]: Done 562 tasks      | elapsed:  3.4min
```

It does not include the total number of tasks, just how many have been completed.

I tried casting `image_src_locations` to a list, hoping that would trigger Parallel to know how many jobs there were, but it didn't seem to make a difference.

Reply:
There is a "--parallelize_verbosity" flag that defines the joblib verbosity (defaulted to a pretty high level of 10, which is why the number of finished tasks is output). I increased this higher but it doesn't give the count of total tasks, like it shows in the joblib docs. I'm not sure what determines when it shows the "Done X tasks out of Y" message. I also tried superficially adding tqdm to get a progress bar but it's crashing when using it with joblib parallel. Might be tricky to add this in, so for now I've just logged at the start of the process how many images are listed in the source file, and one can extrapolate from the existing "jobs done" count. Can look into a better implementation if this is a big issue down the road.

Reply:
Don't worry about this, not worth tearing your hear out about. A log line right before the Parallel call to state something like "Scheduling 10000 tasks..." might be nice, but also optional.